python/lsst/obs/base/filters.py

```diff
@@ -0,0 +1,194 @@
+# This file is part of obs_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Classes to allow obs packages to define the filters used by an Instrument
+and for use by `lsst.afw.image.Filter`, gen2 dataIds, and gen3 Dimensions.
+"""
+
+__all__ = ("FilterDefinition", "FilterDefinitionCollection")
+
+import dataclasses
+import collections.abc
+
+import numpy as np
+
+import lsst.afw.image.utils
+
+
+class FilterDefinitionCollection(collections.abc.Sequence):
+    """An order-preserving collection of FilterDefinitions.
+
+    Parameters
+    ----------
+    filters : sequence
+        The filters in this collection.
+    """
+    class __Singleton:
+        """Store whether the `lsst.afw.image.Filter` singletons have been
+        defined, to prevent subtle errors if different filter collections
+        attempt to define themselves.
+        """
+        def __init__(self):
+            self.defined = None
```

Comment:
I don't understand the need for the extra level of indirection and the singleton class.  What's wrong with something along the lines of:
```
class FilterDefinitionCollection(...):
    ...
    _defined = None

    def defineFilters(self):
        if FilterDefinitionCollection._defined is None:
            FilterDefinitionCollection._defined = self
            ... # actually put the filters into the afw globals
        elif FilterDefinitionCollection._defined is not self:
            raise RuntimeError(...)
        # if _defined is self, do nothing
```

Reply:
Because I'm dumb and forgot how class member work in python. Yes, your version is much better, and still passes the tests.